### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by Leen-odeh3**

### CODE REVIEW

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    private readonly IFixture fixture;
+
+    public WeatherPublisherTest()
+    {
+        fixture = new Fixture().Customize(new AutoMoqCustomization());
+    }
+
+    private WeatherPublisher CreateWeatherDataPublisher()
+    {
+        var mockTextFormatStrategy = new Mock<ITextFormatStrategy>();
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: Make them as scoped variables, the dependencies, also use fixtures to create the objects.

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        var text = fixture.Create<string>();
+        var botConfig = fixture.Create<BotConfiguration>();
+
+        mockTextFormatStrategy.Setup(x => x.GetWeatherData(It.IsAny<string>())).Returns(fixture.Create<Weather>());
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: Don't use `It.IsAny`, we need to make sure we are passing the right object

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+        return new WeatherPublisher(text, mockTextFormatStrategy.Object, botConfig);
+    }
+
+
+    [Fact]
+    public void WeatherDataPublisher_AttachSubscriber()
+    {
+        var sut = CreateWeatherDataPublisher();
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs:37]: as said before, make them variables

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        var mockSubscriber = new Mock<IWeatherSubscriber>();
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: as above

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        var botConfig = fixture.Create<BotConfiguration>();
+
+        mockTextFormatStrategy.Setup(x => x.GetWeatherData(It.IsAny<string>())).Returns(fixture.Create<Weather>());
+
+        return new WeatherPublisher(text, mockTextFormatStrategy.Object, botConfig);
+    }
+
+
+    [Fact]
+    public void WeatherDataPublisher_AttachSubscriber()
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: where is the expected?

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        var sut = CreateWeatherDataPublisher();
+        var mockSubscriber = new Mock<IWeatherSubscriber>();
+
+        sut.Attach(mockSubscriber.Object);
+
+        Assert.Contains(mockSubscriber.Object, sut.Subscribers);
+    }
+
+    [Fact]
+    public void WeatherDataPublisher_DetachSubscriber()
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs:50]: same as above about naming

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        mockSubscriber.Verify(
+            subscriber => subscriber.ProcessWeatherUpdate(It.IsAny<Weather>()),
+            Times.Never
+        );
+    }
+    [Fact]
+    public void InitializeSubscribers_AttachesBotsFromBotConfig()
+    {
+        // Arrange
+        var botConfig = new BotConfiguration
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: use Build with Fixtures

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        // Assert
+        Assert.Contains(botConfig.RainBot, sut.Subscribers);
+        Assert.Contains(botConfig.SnowBot, sut.Subscribers);
+        Assert.Contains(botConfig.SunBot, sut.Subscribers);
+        Assert.Equal(3, sut.Subscribers.Count);
+    }
+
+
+    [Fact]
+    public async Task Notify_CallsProcessWeatherUpdateForAllSubscribers()
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs:97]: same for naming

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        // Arrange
+        var sut = CreateWeatherDataPublisher();
+        var mockSubscriber1 = new Mock<IWeatherSubscriber>();
+        var mockSubscriber2 = new Mock<IWeatherSubscriber>();
+
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: fixtures

File: Weather-Monitoring-And-Reporting-Service_Tests/TextFormatStrategyFactoryTests.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+public class TextFormatStrategyFactoryTests
+{
+    [Theory]
+    [InlineData(null)]
+    [InlineData("")]
+    public void GetTextFormatStrategy_EmptyOrNullString_ThrowsException(string? invalidWeatherDataFilePath)
+    {
+        // Arrange & Act
+        Action act = () => TextFormatStrategyFactory.GetTextFormatStrategy(invalidWeatherDataFilePath);
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/TextFormatStrategyFactoryTests.cs]: use var

File: Weather-Monitoring-And-Reporting-Service_Tests/TextFormatStrategyFactoryTests.cs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+        // Assert
+        act.Should().Throw<WeatherDataFilePathNullException>()
+            .WithMessage("Weather data file path is null or empty.");
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/TextFormatStrategyFactoryTests.cs]: assume we have changed the message, this test will fail and we will need to fix it :) . Find a solution for this

File: Weather-Monitoring-And-Reporting-Service_Tests/TextFormatStrategyFactoryTests.cs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    [Fact]
+    public void GetTextFormatStrategy_UnsupportedFileExtension_ReturnsNull()
+    {
+        string weatherDataFilePath = "x.doc";
+
+        TextFormatStrategyFactory.GetTextFormatStrategy(weatherDataFilePath)
+                             .Should().BeNull();
+    }
+
+    */
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/TextFormatStrategyFactoryTests.cs]: either delete or fix

File: Weather-Monitoring-And-Reporting-Service_Tests/TextFormatStrategyFactoryTests.cs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+    [Fact]
+    public void GetTextFormatStrategy_JsonFileExtension_ReturnsJsonStrategy()
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/TextFormatStrategyFactoryTests.cs]: you forgot Act Arrange Assert comments

File: Weather-Monitoring-And-Reporting-Service_Tests/UnitTest1.cs [hunk 14]
```diff
@@ -0,0 +1,11 @@
+namespace Weather_Monitoring_And_Reporting_Service_Tests
+{
+    public class UnitTest1
+    {
+        [Fact]
+        public void Test1()
+        {
+
+        }
+    }
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/UnitTest1.cs]: remove

File: Weather-Monitoring-And-Reporting-Service_Tests/WeatherBotTests/RainBotTest.cs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+        [Fact]
+        public void ProcessWeatherUpdate_Disabled_MustNotBeActivated()
+        {
+            var weatherData = new Weather { Humidity = 60 };
+            var sut = new RainBot
+            {
+                Enabled = false,
+                HumidityThreshold = 70,
+                Message = "It's raining!"
+            };
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/WeatherBotTests/RainBotTest.cs]: fixtures

File: Weather-Monitoring-And-Reporting-Service_Tests/WeatherBotTests/SunBotTest.cs [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+            Enabled = true,
+            TemperatureThreshold = 30,
+            Message = "It's sunny!"
+        };
+
+        // Act
+        sut.ProcessWeatherUpdate(weatherData);
+
+        // Assert
+        Assert.False(sut.Activated, "The SunBot should not be activated when temperature is lower than the threshold.");
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/WeatherBotTests/SunBotTest.cs:56]: same for here about the messages

**Review by ahmad-m-abbas — CHANGES_REQUESTED** [review]
Fix please

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs (hunk 5 above)
Comment (ahmad-m-abbas) [reply Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: ?

**Review by ahmad-m-abbas — CHANGES_REQUESTED** [review]

File: Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs (hunk 5 above)
Comment (Leen-odeh3) [reply Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: That's Right ?

[Fact]
public void WeatherDataPublisher_AttachSubscriber()
{
    var sut = CreateWeatherDataPublisher();
    var mockSubscriber = _fixture.Create<Mock<IWeatherSubscriber>>();

    sut.Attach(mockSubscriber.Object);

    Assert.Contains(mockSubscriber.Object, sut.Subscribers);

    Assert.True(sut.Subscribers.Contains(mockSubscriber.Object));
}

Comment (ahmad-m-abbas) [reply Weather-Monitoring-And-Reporting-Service_Tests/PublisherTest/WeatherPublisherTest.cs]: I mean something like `WeatherDataPublisher_AttachSubscriber_ShouldWork`

File: Weather-Monitoring-And-Reporting-Service_Tests/StrategiesTest/JsonFormatStrategyTest.cs [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+namespace Weather_Monitoring_And_Reporting_Service_Tests.StrategiesTest;
+
+ public class JsonFormatStrategyTest
+{
+
+    [Fact]
+    public void GetWeatherData_ValidJson_ReturnsWeatherObject()
+    {
+        // Arrange
+        string json = "{\"Temperature\":25,\"Humidity\":50}";
```
Comment (ahmad-m-abbas) [inline Weather-Monitoring-And-Reporting-Service_Tests/StrategiesTest/JsonFormatStrategyTest.cs]: use var

**Review by ahmad-m-abbas — APPROVED** [review]

**Pull request merged**